sql/schemachanger: block data type alter of col used in computed cols

spilchen opened the pull request:
Previously, altering the data type of a column used in a generated column was allowed, causing downstream issues when querying the generated column.

To be consistent with postgres, we will now block any attempt to alter the data type of a column used in a computed column.

Epic: CRDB-25314
Closes: #72457
Release note (bug fix): Attempts to alter the data type of a column used in a computed column expression are now blocked.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/125870)
<!-- Reviewable:end -->

rafiss reviewed (approved):
i just had one optional suggestion to possible make the function simpler

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/125870#-:-O-rUTUmDKvoc8ZBPuwV:b-8n0er8)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @spilchen and @wenyihu6)

___
*[`pkg/sql/catalog/schemaexpr/expr.go` line 644 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/125870#-O-rTp6P4AGXN4rAT2pc:-O-rTp6P4AGXN4rAT2pd:b-5wudl4) ([raw file](https://github.com/cockroachdb/cockroach/blob/9a921b7e2ada7f016e75e08f0aae234b787bba0d/pkg/sql/catalog/schemaexpr/expr.go#L644)):*
> ```Go
> 	}
> 	if referencedCols.Contains(dependentColID) {
> 		return errCallback()
> ```

i think we could keep this simpler by having this function return a `(bool, error)` pair. then rather than calling the error callback here, we just leave it to the caller to return the error it wants to.


<!-- Sent from Reviewable.io -->

spilchen committed:
sql/schemachanger: block data type alter of col used in computed cols

Previously, altering the data type of a column used in a generated
column was allowed, causing downstream issues when querying the
generated column.

To be consistent with postgres, we will now block any attempt to alter
the data type of a column used in a computed column.

The update to event_test.go was needed because it was doing an ALTER
that is now blocked. The test was changed to not reference a column in
the computed expression that we intend to change the data type on.

Epic: CRDB-25314
Closes: #72457
Release note (bug fix): Attempts to alter the data type of a column used
in a computed column expression are now blocked.

Signed-off-by: Matt Spilchen <matt.spilchen@cockroachlabs.com>

spilchen reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/125870)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @rafiss and @wenyihu6)

___
*[`pkg/sql/catalog/schemaexpr/expr.go` line 644 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/125870#-O-rTp6P4AGXN4rAT2pc:-O-utBIrBGCOiqvR64b7:b-fuuqyx) ([raw file](https://github.com/cockroachdb/cockroach/blob/9a921b7e2ada7f016e75e08f0aae234b787bba0d/pkg/sql/catalog/schemaexpr/expr.go#L644)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

i think we could keep this simpler by having this function return a `(bool, error)` pair. then rather than calling the error callback here, we just leave it to the caller to return the error it wants to.
</blockquote></details>

Good suggestion, this is much simpler.


<!-- Sent from Reviewable.io -->

spilchen commented:
tftr!

bors r+
